fsspec/implementations/http.py

```diff
@@ -110,25 +151,84 @@ def ls(self, url, detail=True, **kwargs):
+    ls = sync_wrapper(_ls)
+
+    async def _cat(self, url, chunks=False):
+        if chunks < 1:
+            async with self.session.get(url, **self.kwargs) as r:
+                r.raise_for_status()
+                out = await r.read()
+            return out
+        else:
+            size = await _file_size(url, session=self.session, **self.kwargs)
+            sizes = list(range(0, size, chunks)) + [size]
+            out = [
+                get_range(self.session, url, start, end)
+                for start, end in zip(sizes[:-1], sizes[1:])
+            ]
+            out = await asyncio.gather(*out)
+            return b"".join(out)
+
+    cat = sync_wrapper(_cat)
+
+    async def _mcat(self, paths, recursive=False):
+        paths = self.expand_path(paths, recursive=recursive)
+        out = await asyncio.gather(
+            *[asyncio.ensure_future(self._cat(path), loop=self.loop) for path in paths]
+        )
+        return {k: v for k, v in zip(paths, out)}
+
+    mcat = sync_wrapper(_mcat)
+
+    async def _get_file(self, rpath, lpath, chunk_size=5 * 2 ** 20, chunks=0, **kwargs):
+        kw = self.kwargs.copy()
+        kw.update(kwargs)
+        if chunks < 1:
+            async with self.session.get(rpath, **self.kwargs) as r:
+                r.raise_for_status()
+                with open(lpath, "wb") as fd:
+                    chunk = True
+                    while chunk:
+                        chunk = await r.content.read(chunk_size)
+                        fd.write(chunk)
+        else:
+            open(lpath, "wb").close()
+            size = (await self._info(rpath))["size"]
+            starts = list(range(0, size, chunks)) + [size]
+            [
+                await get_range(self.session, rpath, start, end, file=lpath, **kw)
+                for start, end in zip(starts[:-1], starts[1:])
+            ]
+
+    get_file = sync_wrapper(_get_file)
+
+    async def _get(self, rpath, lpath, recursive=False, **kwargs):
+        from fsspec.implementations.local import make_path_posix
+
+        rpath = self._strip_protocol(rpath)
+        lpath = make_path_posix(lpath)
+        rpaths = self.expand_path(rpath, recursive=recursive)
+        lpaths = other_paths(rpaths, lpath)
+        await asyncio.gather(
+            self._get_file(rpath, lpath, **kwargs)
+            for lpath, rpath in zip(lpaths, rpaths)
+        )
 
-    def mkdirs(self, url):
+    def mkdirs(self, url, **kwargs):
         """Make any intermediate directories to make path writable"""
         raise NotImplementedError
 
-    def exists(self, path):
+    async def _exists(self, path):
         kwargs = self.kwargs.copy()
-        kwargs["stream"] = True
         try:
-            r = self.session.get(path, **kwargs)
-            r.close()
-            return r.ok
+            r = await self.session.get(path, **kwargs)
+            async with r:
+                return r.status < 400
         except requests.HTTPError:
```

Comment:
Does the exception caught here need to change?

Reply:
probably; I should check what aiohttp does when a connection fails completely.

Reply:
probably `aiohttp.ClientError`